File: apps/web/src/lib/api/campaigns.ts
"use client";

import { api } from "./client";

export type CampaignKind = "widget_trigger" | "broadcast";
export type CampaignStatus = "draft" | "active" | "paused";

export interface ApiCampaign {
  id: string;
  agent_id: string;
  kind: CampaignKind;
  name: string;
  message: string;
  trigger_config: { delay_seconds?: number; url_pattern?: string };
  status: CampaignStatus;
  created_at: string;
  updated_at: string;
}

export function listCampaigns(agentId?: string) {
  const qs = agentId ? `?agent_id=${agentId}` : "";
  return api<ApiCampaign[]>(`/v1/campaigns${qs}`, { orgScoped: true });
}

export function createCampaign(body: {
  agent_id: string;
  kind?: CampaignKind;
  name: string;
  message: string;
  trigger_config?: Record<string, unknown>;
  status?: CampaignStatus;
}) {
  return api<ApiCampaign>("/v1/campaigns", { method: "POST", orgScoped: true, body });
}

export function updateCampaign(id: string, body: Record<string, unknown>) {
  return api<ApiCampaign>(`/v1/campaigns/${id}`, { method: "PATCH", orgScoped: true, body });
}

export function deleteCampaign(id: string) {
  return api<void>(`/v1/campaigns/${id}`, { method: "DELETE", orgScoped: true });
}

File: apps/api/app/models/campaigns.py
"""Campaigns — proactive outbound messages.

Two very different things share the name, and only one of them is safe to ship as-is:

- ``widget_trigger``: a message the embedded widget shows unprompted after a client-side
  condition (time on page, URL match). No external channel, no consent question — the
  visitor is already on the site.
- ``broadcast``: sending to many contacts on WhatsApp/etc. That needs consent tracking, an
  unsubscribe path, rate-limited batch sending, and (outside WhatsApp's 24-hour window)
  approved templates. Until that exists, broadcasts are **draft-only and cannot send** —
  see `docs/DECISIONS.md` ADR-039.
"""

from __future__ import annotations

import uuid
from typing import Any

from sqlalchemy import ForeignKey, String, Text
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base, TimestampMixin, UUIDPrimaryKey

KINDS = ("widget_trigger", "broadcast")
STATUSES = ("draft", "active", "paused")


class Campaign(Base, UUIDPrimaryKey, TimestampMixin):
    __tablename__ = "campaigns"

    organization_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("organizations.id", ondelete="CASCADE"), index=True
    )
    agent_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("agents.id", ondelete="CASCADE"), index=True)
    kind: Mapped[str] = mapped_column(String(16), nullable=False)  # widget_trigger | broadcast
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    message: Mapped[str] = mapped_column(Text, nullable=False)
    #: widget_trigger: {"delay_seconds": int, "url_pattern": str}
    trigger_config: Mapped[dict[str, Any]] = mapped_column(JSONB, default=dict, nullable=False)
    status: Mapped[str] = mapped_column(String(16), default="draft", nullable=False)
